# Homework 5 - Logistic Regression

## Your Name Here

Student Name:

Student UT EID:

---

Partner Name:

Partner UT EID:

---

Date Created:

Date Last Modified:

---

Totoal Points 20. 

In [293]:
# Standard Headers
# You are welcome to add additional headers here if you wish
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, classification_report

# Enable inline mode for matplotlib so that Jupyter displays graphs
%matplotlib inline

## Heart Dataset 

In this Assignment we will work with some patients dataset. 

We have access to 303 patients data. The features are listed below. 

In [294]:
# Your code here
heart_df = pd.read_csv("/Users/dylanlam/Documents/GitHub/data_science_practice_and_skills/datasets/Heart.csv")
heart_df

Age  Sex     ChestPain  RestBP  Chol  Fbs  RestECG  MaxHR  ExAng  \
0     63    1       typical     145   233    1        2    150      0   
1     67    1  asymptomatic     160   286    0        2    108      1   
2     67    1  asymptomatic     120   229    0        2    129      1   
3     37    1    nonanginal     130   250    0        0    187      0   
4     41    0    nontypical     130   204    0        2    172      0   
..   ...  ...           ...     ...   ...  ...      ...    ...    ...   
298   45    1       typical     110   264    0        0    132      0   
299   68    1  asymptomatic     144   193    1        0    141      0   
300   57    1  asymptomatic     130   131    0        0    115      1   
301   57    0    nontypical     130   236    0        2    174      0   
302   38    1    nonanginal     138   175    0        0    173      0   

     Oldpeak  Slope   Ca        Thal Target  
0        2.3      3  0.0       fixed     No  
1        1.5      2  3.0      normal    Yes  
2        2.6      2  2.0  reversable    Yes  
3        3.5      3  0.0      normal     No  
4        1.4      1  0.0      normal     No  
..       ...    ...  ...         ...    ...  
298      1.2      2  0.0  reversable    Yes  
299      3.4      2  2.0  reversable    Yes  
300      1.2      2  1.0  reversable    Yes  
301      0.0      2  1.0      normal    Yes  
302      0.0      1  NaN      normal     No  

[303 rows x 14 columns]

**Age:** The person’s age in years

**Sex:** The person’s sex (1 = male, 0 = female)

**ChestPain:** chest pain type

* Value 0: asymptomatic
* Value 1: atypical angina
* Value 2: non-anginal pain
* Value 3: typical angina

**RestBP:** The person’s resting blood pressure (mm Hg on admission to the hospital)

**Chol:** The person’s cholesterol measurement in mg/dl

**Fbs:** The person’s fasting blood sugar (> 120 mg/dl, 1 = true; 0 = false)
restecg: resting electrocardiographic results

* Value 0: showing probable or definite left ventricular hypertrophy by Estes’ criteria
* Value 1: normal
* Value 2: having ST-T wave abnormality (T wave inversions and/or ST elevation or depression of > 0.05 mV)

**RestECG:** The person’s maximum heart rate achieved

**MaxHR:** Exercise induced angina (1 = yes; 0 = no)

**Oldpeak:** ST depression induced by exercise relative to rest (‘ST’ relates to positions on the ECG plot. See more here)

**Slope:** the slope of the peak exercise ST segment — 0: downsloping; 1: flat; 2: upsloping

* 0: downsloping; 
* 1: flat; 
* 2: upsloping

**Ca:** The number of major vessels (0–3)

**Thal:** A blood disorder called thalassemia Value 0: NULL (dropped from the dataset previously

* Value 1: fixed defect (no blood flow in some part of the heart)
* Value 2: normal blood flow
* Value 3: reversible defect (a blood flow is observed but it is not normal)

**Target:** Heart disease (1 = no, 0= yes)

# Task - 1 (5 points)
We want to use logistic regerssion to perdict if the patients will have heart problems or not. The column "Target" in our datasets includes data about heart diseases. If the patient had heart disease we have a 1 and if not a zero. 

Prepare your data set for predicting heart disease ("Target" column) out of 3 features:

* Age of the patient (Column **"Age"**)
* Gender of the patient (male or female - Column **"Sex"**)
* Cholestrol level of the patient (Column **"Chol"**) 


Split your data into 80% traning data and 20% test data. 

In [295]:
X = heart_df[['Age', 'Sex', 'Chol']]
y = heart_df['Target']

# Split the data into training and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


# Task 2 - (10 points)

Generate a logistic regression model using your training data. 

Print out the accuracy if your model. 

In [296]:
# Add your code Here! 
model = LogisticRegression()
model.fit(X_train, y_train)

y_pred = model.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)


Accuracy: 0.5737704918032787


# Task 3 - (5 points)


Generate the classification report for your Logistic regresion model and interpret your results regarding precision, recall and f1-score. 


In [297]:
report = classification_report(y_test, y_pred)

print(report)

              precision    recall  f1-score   support

          No       0.55      0.62      0.58        29
         Yes       0.61      0.53      0.57        32

    accuracy                           0.57        61
   macro avg       0.58      0.58      0.57        61
weighted avg       0.58      0.57      0.57        61



**Precision:**

- For the "No" class, the precision is 0.55, meaning that among all the instances predicted as "No", 55% were actually "No".
- For the "Yes" class, the precision is 0.61, indicating that among all the instances predicted as "Yes", 61% were actually "Yes".

**Recall:**

- For the "No" class, the recall is 0.62, indicating that the model correctly identified 62% of all the actual "No" instances.
- For the "Yes" class, the recall is 0.53, indicating that the model correctly identified 53% of all the actual "Yes" instances.

**F1-score:**

- For the "No" class, the F1-score is 0.58, which is the harmonic mean of precision and recall for the "No" class.
- For the "Yes" class, the F1-score is 0.57, which is the harmonic mean of precision and recall for the "Yes" class.

**Support:**

- The support indicates the number of actual occurrences of each class in the test dataset. For example, there were 29 instances of "No" and 32 instances of "Yes" in the test dataset.

**Accuracy:**

- The overall accuracy of the model is 0.57, which means that the model correctly predicted the class labels for 57% of the instances in the test dataset.

**Macro Avg:**

- The macro average precision, recall, and F1-score are calculated by averaging the precision, recall, and F1-score across all classes, without considering the class imbalance. Here, the macro average F1-score is 0.57.

**Weighted Avg:**

- The weighted average precision, recall, and F1-score are calculated by averaging the precision, recall, and F1-score across all classes, with each class's score weighted by its support. Here, the weighted average F1-score is 0.57.

# Task 4 - Optional Task (2 extra points)


Which other feature of your data can you use to improve your prediction? Build other models and improve your prediction. 


**ChestPain:** chest pain type

* Value 0: asymptomatic
* Value 1: atypical angina
* Value 2: non-anginal pain
* Value 3: typical angina

In [298]:
# One-hot encoding ChestPain
heart_df.loc[heart_df['ChestPain'] == 'asymptomatic', 'ChestPain'] = 0
heart_df.loc[heart_df['ChestPain'] == 'nontypical', 'ChestPain'] = 1
heart_df.loc[heart_df['ChestPain'] == 'nonanginal', 'ChestPain'] = 2
heart_df.loc[heart_df['ChestPain'] == 'typical', 'ChestPain'] = 3
heart_df[['ChestPain']]

ChestPain
0           3
1           0
2           0
3           2
4           1
..        ...
298         3
299         0
300         0
301         1
302         2

[303 rows x 1 columns]

In [299]:
X2 = heart_df[['Age', 'Sex', 'Chol', 'ChestPain']]
X2_train, X2_test, y_train, y_test = train_test_split(X2, y, test_size = .2, random_state = 42)
model2 = LogisticRegression()
model2.fit(X2_train, y_train)
y_pred = model2.predict(X2_test)
print('Accuracy:', accuracy_score(y_test, y_pred))
print(classification_report(y_test, y_pred))

Accuracy: 0.6885245901639344
              precision    recall  f1-score   support

          No       0.68      0.66      0.67        29
         Yes       0.70      0.72      0.71        32

    accuracy                           0.69        61
   macro avg       0.69      0.69      0.69        61
weighted avg       0.69      0.69      0.69        61



In [300]:
X3 = heart_df[['Age', 'Sex', 'Chol', 'ChestPain', 'MaxHR']]
X3_train, X3_test, y_train, y_test = train_test_split(X3, y, test_size = .2, random_state = 42)
model3 = LogisticRegression()
model3.fit(X3_train, y_train)
y_pred = model3.predict(X3_test)
print('Accuracy:', accuracy_score(y_test, y_pred))
print(classification_report(y_test, y_pred))

Accuracy: 0.7868852459016393
              precision    recall  f1-score   support

          No       0.77      0.79      0.78        29
         Yes       0.81      0.78      0.79        32

    accuracy                           0.79        61
   macro avg       0.79      0.79      0.79        61
weighted avg       0.79      0.79      0.79        61



In [301]:
X4 = heart_df[['Age', 'Sex', 'Chol', 'ChestPain','MaxHR', 'Oldpeak', 'RestBP', 'RestECG', 'Slope', 'Fbs', 'ExAng']]
X4_train, X4_test, y_train, y_test = train_test_split(X4, y, test_size = .2, random_state = 42)
model4 = LogisticRegression()
model4.fit(X4_train, y_train)
y_pred = model4.predict(X4_test)
print('Accuracy:', accuracy_score(y_test, y_pred))
print(classification_report(y_test, y_pred))

Accuracy: 0.8032786885245902
              precision    recall  f1-score   support

          No       0.76      0.86      0.81        29
         Yes       0.86      0.75      0.80        32

    accuracy                           0.80        61
   macro avg       0.81      0.81      0.80        61
weighted avg       0.81      0.80      0.80        61



/Users/dylanlam/miniconda3/envs/myenv/lib/python3.8/site-packages/sklearn/linear_model/_logistic.py:458: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
